filter_test.go

```diff
@@ -0,0 +1,241 @@
+package main
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func Test_FilterParse(t *testing.T) {
+	// fixtures
+	presence := &Presence{Attr: "present"}
+	equality := &Equality{Attr: "eq", Value: "eqval"}
+
+	tests := []struct {
+		name         string
+		filter       string
+		expectedNode FilterNode
+	}{
+		{
+			name:         "presence",
+			filter:       "(present=*)",
+			expectedNode: presence,
+		},
+		{
+			name:         "equality",
+			filter:       "(eq=eqval)",
+			expectedNode: equality,
+		},
+		{
+			name:         "and one",
+			filter:       "(&(eq=eqval))",
```

Comment:
with all due respect - that is one hack of weird syntax....

Reply:
with all due respect - this is LDAP ma'am. Weird is only the start.